### ISSUE TITLE
[bug] 'compiled-objects-have-debug-symbols' false positive on mach-o files that have been passed through 'strip'

### ISSUE THREAD

**Issue by jameslamb** [jameslamb/pydistcheck#235]
### What did you expect to happen?

Expected `compiled-objects-have-debug-symbols` check to only find debug symbols if a mach-o binary contains them.

### What actually happened?

Over in LightGBM, observed that when upgrading `scikit-build-core` for a shared library from v0.4.4 to 0.9.3, `pydistcheck` started raising this on the resulting library `lib_lightgbm.dylib`.

```text
1. [compiled-objects-have-debug-symbols] Found compiled object containing debug symbols. For details, extract the distribution contents and run 'nm -a "lightgbm/lib/lib_lightgbm.dylib"'.
errors found while checking: 1
```

But ONLY that macOS library... not the equivalent files on linux (`lib_lightgbm.so`) or windows (`lib_lightgbm.dll`).

### How can someone else reproduce this problem?

I mention that `scikit-build-core` version because in that range, `scikit-build-core` added the ability to strip the binaries it produced.

https://github.com/scikit-build/scikit-build-core/blob/2994db5d993f0cb252dbeebfac7d66677b670ed8/README.md#L278-279

I built that library (`lib_lightgbm.dylib`) on an M2 mac, in a build driven by `scikit-build-core`, like this:

```shell
git clone \
    --recursive \
    git@github.com:microsoft/LightGBM.git \
    /tmp/lightgbm

cd /tmp/lightgbm
sh build-python.sh bdist_wheel
unzip -d ./tmp ./dist/*.whl
nm ./lightgbm/lib/lib_lightgbm.dylib > o.txt
nm -a ./lightgbm/lib/lib_lightgbm.dylib > a.txt
```

Then diffed those files. Only saw 1 difference.

<img width="1100" alt="image" src="https://github.com/jameslamb/pydistcheck/assets/7608904/41f78254-67c0-4a39-8ea4-2e092c30f2fc">

The results of `nm -a` includes this:

```text
0000000005614542 - 00 0000   OPT radr://5614542
```

It seems that maybe this is added by `strip`?

I found this repo, https://github.com/nico/lssym, which says

```text
(Compare with the output of `nm lssym` and `dsymutil -s lssym` in both cases.
Note that `strip` adds a symbol that looks like

    radr://5614542    N_STAB 3c    n_sect 000 n_desc 0x0000    n_value 0x5614542

I wonder what that bug is. `nm` doesn't list it because `nm` only lists symbols,
and this is a stabs debug info entry which `nm` only shows if you pass `-a`.)
```

I strongly suspect that's what's happening here! Not sure what the best fix is though for `pydistcheck`, will need to think through that.

### What version of `pydistcheck` are you using?

0.6.0

### Notes

Thanks @nico for putting up https://github.com/nico/lssym! It was a great clue to help me investigate this!!

**Comment by agriyakhetarpal** [comment]
Looks like this is a special case that might be possible to ignore safely, because Xcode's `strip` prefers to keep weak symbols, see: https://x.com/yaakov_h/status/478656504162029568/photo/1

The corresponding Apple Open Source pages have unfortunately been wiped into oblivion:

**Comment by agriyakhetarpal** [comment]
Looks like this is a special, corner case that might be possible to ignore safely, because Xcode's `strip` prefers to keep weak symbols like , see: https://x.com/yaakov_h/status/478656504162029568/photo/1

Unfortunately, the corresponding Apple Open Source pages have been wiped into oblivion. I tried
- https://opensource.apple.com/source/cctools/cctools-822/misc/strip.c
- https://opensource.apple.com/source/cctools/cctools-973.0.1/misc/strip.c.auto.html
- https://opensource.apple.com/source/cctools/cctools-751/misc/strip.c.auto.html

and I naïvely thought I could look them up on archival services such as https://archive.is/opensource.apple.com or on the Internet Archive's Wayback Machine, but to no avail – it is apparent that Apple has probably worked aggressively to stop the spread of this knowledge.

It is a remnant of the linker, though, so maybe using alternative linkers such as [Zld](https://steipete.com/posts/zld-a-faster-linker/) will avoid this problem.

Even on this page: https://docs.angr.io/projects/cle/en/v9.2.70/_modules/cle/backends/macho/symbol.html, the sentence seems incomplete:

> `# The addr isn't really an address, but a magic value that is used to indicate that the symbol is`

I don't claim any conspiracies here, though. 😃

IDA 7.6 seems to ignore it as well: https://docs.hex-rays.com/release-notes/7_6#file-formats

**Comment by agriyakhetarpal** [comment]
I managed to find a link that still works: https://archive.is/7voAu

**Comment by jameslamb** [comment]
Wow AMAZING investigation!!! Thank you so much for looking into that.

Are you interested in submitting a PR that ignores this specific symbol when `nm` / `llvm-nm` are use on mach-o` files here? If not, no worries, I'd be happy to attempt it.

I think it'd involve filtering that out around here:

https://github.com/jameslamb/pydistcheck/blob/85b8af32b87eb8d9def9fe318381a1704be41aba/src/pydistcheck/shared_lib_utils.py#L65-L68

with a code comment pointing back to this issue to explain why

**Comment by agriyakhetarpal** [comment]
Thank you for the pointer! Yes, I'm interested in contributing and will put together a PR this week :D
